### ISSUE TITLE
Problem text log of selector switch

### ISSUE THREAD

**Issue by rbroeders** [domoticz/domoticz#4143]
Hi,

I noticed that the behaviour of the text log is changed to avoid dupplicate messages in the log. But this makes the log for hardware buttons which uses a selector switch widget unusable as it only logs a state change now. So when I have a button which is mainly used with a single click I will never see a log message. I will only see a single click when previously someone double clicked.
The result is that the current log is misleading because I can't see anymore when someone used a button. I really want to have a good logging back.

Can the selector switch be added to the list of exceptions?
Or maybe a setting on the widget to allow duplicate log entries for that widget, this will give more flexibility.

See also https://www.domoticz.com/forum/viewtopic.php?f=6&t=32569&sid=8b6b1f52a2c885ef7855e9503b439bde

**Comment by laoranz** [comment]
Hi,

I have the same problem with another detector.
It's a motion detector seen as a switchType onOff: Type=32, SubType=8, SwitchType=0

Only 1 message is sent : "Motion", so after this commit https://github.com/domoticz/domoticz/commit/e07e74975, no more messages are logged.

As rbroeders said, may be provide a selector in the widget setting to force the log if value change or in all cases. As this has been done for sending data to influxdb.

**Comment by gizmocuz** [comment]
@laoranz , does this sensor not switch itself off after some time ? If not, edit the switch and specify a auto off delay (let's day 300 seconds (5 minutes))

**Comment by laoranz** [comment]
@gizmocuz, no, It is a motion detector. It only transmits 1 type of message upon detection and no more after.
I don't have the possibility to defined an auto off in the device setting. This device was created automatically by domoticz.

![image](https://user-images.githubusercontent.com/25210030/82950155-1e993a00-9fa5-11ea-88ad-b7a5c3a37b98.png)

I tried to force the value of the device to off with a dzEvent script, but this generated an alarm (opening of the motion detector ^^)

By modifying the switchType _OnOff_ by a _PushOn_ directly in the database, I have logs again:

![image](https://user-images.githubusercontent.com/25210030/82950621-e1817780-9fa5-11ea-9b70-a8c52398989c.png)

But it isn't really a solution.

**Comment by gizmocuz** [comment]
@laoranz , I don't know what theme you are using, but if you use the default theme, and the switch type = motion, you should have a On/Off delay

![image](https://user-images.githubusercontent.com/2265299/82991072-68ae0a00-9ffd-11ea-9e5d-463908253ab2.png)

**Comment by laoranz** [comment]
@gizmocuz, even with the default theme, i don't have these options.
It seems it's because it's a security device. Some comments in html indicate that these options are hidden
`<!-- ngIf: ::!vm.isSecurityDevice() -->`

This device was automatically generated by RFXcom hardware.

To resolve this issue, maybe add a checkbox to bypass the duplication check before writing to the log ? Or add in the exception list some types of devices .

**Comment by rbroeders** [comment]
The checkbox would be great :)

**Comment by gizmocuz** [comment]
@laoranz , You would expect this device to return to the state 'Normal' or 'Normal Delayed' after it detected motion (if there was no motion anymore)
I added the pTypeSecurity1 to the list of exceptions in beta 12102

@rbroeders , No, I will not add a checkbox for a few devices that would need it as this will give to many options and will confuse users. There are already exceptions for PushOn/Off, doorbell etc..
Maybe a few more need to be added as mentioned above
Could you post a screenshot of your selector switch ?

**Comment by rbroeders** [comment]
@gizmocuz Adding the selector switch to the exception list would really help me.
Below is one of the selector switches, but the same applies to all other hardware switches which have a selector switch as widget. So all who support click, double click, triple click, etc.

![C3FAC1F1-5677-43E0-B3FC-F80F0363DCFD](https://user-images.githubusercontent.com/45007717/83106055-167fee00-a0bc-11ea-9391-a2ca1aefca2c.jpeg)
![BE0B82DE-CDC0-4767-BD20-761534D6D573](https://user-images.githubusercontent.com/45007717/83106064-18e24800-a0bc-11ea-94b8-3cf5e96a7b93.png)
![F6876BBF-8D95-47F8-B946-DB2DD57833B3](https://user-images.githubusercontent.com/45007717/83106069-1b44a200-a0bc-11ea-8874-2abd037896d1.png)

**Comment by gizmocuz** [comment]
@rbroeders , the whole reason we only log changes is that previously people use (and still using) bad scripts that adds a log entry every second (if status = off, then update switch with off, instead of checking if there is actually a different state in the switch and if so, update it)
I can not add this bypass as one of the problems was with selector switches
There is probably a workaround that you can make, add another level that does nothing (label normal or something) 
Then make a dzVents script that when you activate any selector state, it will set this dummy state after xx seconds

**Comment by rbroeders** [comment]
@gizmocuz I fully understand the reason of this change and yes the workaround you propose is possible but I don't see it as a real solution. Can't do it in a script in a generc way because there too much differences. I have 33 selector switches with different hehaviours:
- Some set a level or a state so need to keep the selected level. For these I have no problem with the change.
- For each hardware switch I need to make an exception for the long press, it shouldn't go back to off but wait for the press release. Different hardware switches use different levels for the long press. And some are actially two switches with one widget with a left long press, right long press and both long press.

I get your point about "Allow duplicate log entries", although I don't think it's that confusing I accept your choice.

I guess the bad scripts all use the dummy device hardware, maybe it's a good solution to only block duplicates for dummy selector switches and allow them from Zigate, Xiaomi, conbee, zigbee2mqtt, etc.

**Comment by gizmocuz** [comment]
@rbroeders , the 'Selector Switch' is already a 'strange duck' in Domoticz , and 33 of them ?
What are you doing with all of these ? That seems crazy ;)
Personally triple clicking a button that will cause another light to go on that if you double click it, or long hold it, or single click it... it does not seem to have a very good acceptance factor for other household members at all.
Also when you can translate a single switch (dummy with a script for instance) to that triple click to turn something on, that would be a much better implementation on your dashboard.
Having 10 selector options and needing to select the dropdown to choose a action is hard on small devices, but also takes a lot of time.
Hiding this switch (by adding a $ sign at the beginning of the name) and using dummy (push) switches seems the way to go.
A user with 33 selector switches I also consider not an average users, and I suppose creating a dzVents script (or blockly) as a workaround should not be hard to make

**Comment by rbroeders** [comment]
@gizmocuz Please don't judge the amount of selector switches I have ;) Like I said most of them are zigbee hardware buttons and Xiaomi and zigate creates selector switches for them. The new Xiaomi Aquara remote creates 6 selector switches for one device, so that goes fast and got three of them. I also agree with triple click, I don't use it but the device supports it so it has a level.
I also agree that they are not user friendly in the GUI, but they are hardware switches so I don't use them in the GUI. I use the real button. Basically the logging is the only important part of the widget.
If you want to know, double click to switch another light is something I don't do. I agree that doesn't help in user experience. But click for on/off, double click to rotate through a few white color settings and long press to dim up/down until a click release is something that works fine.

Hiding the switch and use push buttons is not the way to go, the bulbs have their own widgets and those are used to control the lights from the gui. I do hide the switches, just want to have a good log.

But basically this all is off-topic. Yes I can make a workaround like you proposed but that's hard to maintain because I have to take the long press into consideration and different buttons use different levels for that. I can also make a new log file from dzVents which seems a better workaround but it's not user friendly.
 
But that's not the point. The point is we had a good logging and now it's useless for all hardware buttons. At this moment it's better to remove the log for the widget because what's in it is just not true. And I think there are better solutions to prevent the problem of bad scripts, like the proposed checkbox, depending on hardware type or even a max of 250 log messages per selector switch.

**Comment by gizmocuz** [comment]
Maybe the Xiaomi should not use selector switches but regular switches. You could ask the author to change this.
It's like some ZWave switches, but they are only added when they are actually used. If nobody is triple long clicking on the button, there will be no such switch added to the system.

**Comment by rbroeders** [comment]
I'll talk to the the zigate author about a possible solution. Don't see your proposal happening, that would mean a very big change for a lot of people. And sure it's possible to have a pushOn button for the every different click and a switch button for a long click/release. It would give  a lot of extra devices. The 6 button Aqara Opple switch then would create 18 pushOn buttons and 6 switches...

Still think this is the world upside down. Logging should be a basic functionality, but because some people have bad scripts the whole world needs to change...

I would be happy to think along for a different solution on the log problem but I see you're not willing to change it.

**Comment by pipiche38** [comment]
@gizmocuz , I discover this big change on the behaviour of Domoticz Switches, and on top of what is reported by @rbroeders , does that mean that updating for instance a switch only with Battery and Signal value will result in a NO update .

I found very strange that because user are having bad scripting that a change is made on the main application to avoid end results due to bad usage !

**Comment by arjanvanleent** [comment]
I use Zigbee2mqtt and I see normal logs and this is a Xiaomi switch

![image](https://user-images.githubusercontent.com/17556892/83323804-6cda6180-a261-11ea-8b27-da8a8bf69354.png)

**Comment by rbroeders** [comment]
Looks like zigbee2mqtt switches to off right after the event. The zigate and Xioami plugin don't do the off.
Can you show a part of the log when you longpress/release a button?
Thanks.

**Comment by arjanvanleent** [comment]
Sure here it is

![image](https://user-images.githubusercontent.com/17556892/83324148-7c5aaa00-a263-11ea-90b3-ee0899c4c829.png)

**Comment by rbroeders** [comment]
Thanks, I wouldn't expect the off right after the long click.

**Comment by deennoo** [comment]
On my side, i use domoticz since 5 years now, i have to use 433 then zwave then zigbee device.

For device who use selector and on/off status like switch "not Motion or Door Sensor" the ability to make script/event starting like : 


```
if devicechanged ('devicename') then
 commandarray ['slavedevicename'] : "toggle"
```


Was a great and wonderfull feature of easyness using scriptingon LUA, user don't have to care which position was press, Of course have multiple state event luncher is great too for complex event.

i thing we are going a big step backward since this change, in our days every thing is made to make thing easy.

Please restore this as it was

**Comment by pipiche38** [comment]
> I use Zigbee2mqtt and I see normal logs and this is a Xiaomi switch
> 
> ![image](https://user-images.githubusercontent.com/17556892/83323804-6cda6180-a261-11ea-8b27-da8a8bf69354.png)

@arjanwaard I'm sorry you don't have a correct log, you have a Off straight after the real event, which is not correct. 
What I see is that zigbee2mqttt had tweak their plugin in order to make it working with Domoticz, while this doesn't represent the device event !

**Comment by laoranz** [comment]
> 
> 
> @laoranz , You would expect this device to return to the state 'Normal' or 'Normal Delayed' after it detected motion (if there was no motion anymore)
> I added the pTypeSecurity1 to the list of exceptions in beta 12102

@gizmocuz, thank you for this correction, I think it is not necessary to have a 'Normal' state for this device. This detector creates an event, not really a state. But if it's inperative, 'Normal Delayed' since be a good solution.
But this patch is only for a particular case.

Checking for duplicates was a good solution for dealing with bad scripts, but it seems to generate others problems for logs.
As this check was added due to bad scripts. Is it possible to apply it only to scipts? Of course, if it is possible to know the origin of the action.

**Comment by Smanar** [comment]
Sorry If I m wrong but I m not sure I have understand all the issue, and I havent installed the last version to be sure, too much constraint.

The log only log the last state, I m agree with that.
But if a python script update 2 time the same widget with the same state,
>Devices[8].Update(nValue=10, sValue=10)
Devices[8].Update(nValue=10, sValue=10)

Domoticz will log only 1 command, but what happen for lua script ? It trigger 1 or 2 time ?
Because I m making like zigbee2mqtt in my plugin, I set the switch to off after the command. But it's not always the better solution.

For exemple you can have, only 1 value for long press, and only 1 for the release, but using periodic value during a long press permit lot of thing, for exemple, you can make light variation if you receive a "long press" event every 0.5s , but not if you use the "long press released" event.

And not possible force a "off" value after every "long press event" if this one appears every 0.5s
And worse, it can be usefull to memorise the last state on some situation, better than setting it to "off".

Edit:
Seriously after reflexion, I realy think forcing the "off" is the worst solution. And now I understand why it s not used in the kodi hardware for exemple.

**Comment by rbroeders** [comment]
@Smanar all the event triggers for scripts do happen. Only issue is the logging of the event.
For example, I have a Xioami button connected to ZiGate to turn on/off a light with a single click. Also have the double click and long press/release implemented to go through predefined setpoints and dim up/down. But don’t use that much. So normally only the single click is used and that works fine. 
ZiGate does not force the widget to off. Which is correct in my opinion, the button doesn’t go off and the widget just shows the last event... when I look in my log the last click was weeks ago although it’s used more frequently. Just because the last log entry was a click, new click events aren’t logged again.
I agree forcing an off after an event is a bad idea. It will be a little bit better to force it to off just before the event. In that way the last real event remains on the widget. But then the log is filled with false off events which actually never happened.

**Comment by Smanar** [comment]
Ha ok, so sorry, forget me ^^

But in this situation I don't see the problem, except for debugging, but for normal use, there is not difference ? All is working fine with zigate too, we don't need to force the "off" event to make script working ?

**Comment by rbroeders** [comment]
No, the issues is just about logging. Script run fine.
Although the workaround with a forcing off will have effect on scripts.

**Comment by Smanar** [comment]
Yes, and I m agree with you, I m using it too on my side, but the forcing is just cosmetic (To don't have a switch locked on a state visually) but it's delayed and under condition.

BTW sorry for the interruption and thx for information.

**Comment by pipiche38** [comment]
> Yes, and I m agree with you, I m using it too on my side, but the forcing is just cosmetic (To don't have a switch locked on a state visually) but it's delayed and under condition.
> 
> BTW sorry for the interruption and thx for information.

Forcing will generate an event and if you have scripts based on devicechanged as @deennoo this will not work anymore.

```
if devicechanged ('devicename') then
 commandarray ['slavedevicename'] : "toggle"
```

**Comment by rbroeders** [comment]
@gizmocuz I think there could be a nice solution. Can you make a new parameter in the device options section of the Python plugin API? I'm thinking of "allowDuplicateLog" and when allowDuplicateLog=true then bypass the duplicate check. When allowDuplicateLog=false or is missing the behavior would be the current one. This way plugin creators who know what they are doing could still allow duplicate log entries and bad scripts will be safe.
Also no extra complexity for end users.

I think this would be a great workaround.

**Comment by gizmocuz** [comment]
The idea is to only add the sensor do Domoticz when it has been used
So if you are not 6 times pressing the button you don't get that sensor in Domoticz
That hardware should not add this hardware just because it is possible

**Comment by rbroeders** [comment]
But all the plugins already use selector switches... this way with minimal effort we can get the previous functionality back and your problem with bad scripting is also solved.

Personally I don’t really like devices to be added when someone does another kind of click. I rather have all the possibilities from the start. I can remove unused devices from the GUI and remove unused states from the selector switch (although these don’t bother me)

I think my proposal would be a minor change and all plugin developers can use it if they want. Not only for selector switches but for all widgets. Possibly you could replace your white list by the default setting of this option per device type.

**Comment by deennoo** [comment]
> The idea is to only add the sensor do Domoticz when it has been used
> So if you are not 6 times pressing the button you don't get that sensor in Domoticz
> That hardware should not add this hardware just because it is possible

Don't you think that don't show this aviability, make blind any possibility ?

I'm with you about 3/4 click on a Switch, easy for the script autor but never use form all familly, but please let us choose if we need it or not.

**Comment by pipiche38** [comment]
@rbroeders suggestions is to use the existing Options DeviceStatus field and on top of existing attributes to add AllowDuplicateLogs ( equal to 'true' or 'false' ).
Of course if this attribute doesn't exist then Domoticz will behave accordingly to the today behaviour (no duplicate in device log)

Existing Options attributes in Domoticz DeviceStatus
LevelNames
LevelActions
SelectorStyle
LevelOffHidden

Then it will up to each plugin developer to populate the attribute in order to get allow to write duplicate in the logs as it was possible in the previous version.

I do consider as a fair request .

**Comment by rbroeders** [comment]
@gizmocuz You didn't really respond to the solution with an 'allowDuplicateLog' option through the Python plugin API. Would this be a workable solution for you? It would prevent workaround where plugins or scripts need to set the selectors to an off state.
